File: sdk/python/iploop/retry.py
"""Retry logic with IP rotation."""

import time
import uuid
import logging

logger = logging.getLogger("iploop")

RETRYABLE_STATUS = {403, 407, 429, 500, 502, 503, 504}


def should_retry(exc=None, status_code=None):
    if status_code and status_code in RETRYABLE_STATUS:
        return True
    if exc:
        from requests.exceptions import (ConnectionError, Timeout, ProxyError)
        return isinstance(exc, (ConnectionError, Timeout, ProxyError))
    return False


def new_session_id():
    return uuid.uuid4().hex[:16]


def retry_request(func, retries=3, delay=1.0):
    """Execute func with retries. func receives attempt number."""
    last_exc = None
    for attempt in range(retries):
        try:
            resp = func(attempt)
            if hasattr(resp, 'status_code') and should_retry(status_code=resp.status_code):
                logger.debug("Retry %d/%d: status %d", attempt + 1, retries, resp.status_code)
                if attempt < retries - 1:
                    time.sleep(delay * (attempt + 1))
                    continue
            return resp
        except Exception as e:
            last_exc = e
            if should_retry(exc=e) and attempt < retries - 1:
                logger.debug("Retry %d/%d: %s", attempt + 1, retries, e)
                time.sleep(delay * (attempt + 1))
            elif attempt == retries - 1:
                break
            else:
                raise
    raise last_exc

File: sdk/node/src/retry.js
const crypto = require('crypto');
const RETRYABLE_STATUS = new Set([403, 407, 429, 500, 502, 503, 504]);
function newSessionId() { return crypto.randomBytes(8).toString('hex'); }
function shouldRetry(err, statusCode) {
  if (statusCode && RETRYABLE_STATUS.has(statusCode)) return true;
  if (err && (err.code === 'ECONNRESET' || err.code === 'ECONNREFUSED' || err.code === 'ETIMEDOUT' || err.code === 'EPIPE')) return true;
  return false;
}
async function retryRequest(fn, retries = 3, delay = 1000) {
  let lastErr;
  for (let i = 0; i < retries; i++) {
    try {
      const result = await fn(i);
      if (result && result.status && shouldRetry(null, result.status)) {
        if (i < retries - 1) { await sleep(delay * (i + 1)); continue; }
      }
      return result;
    } catch (e) {
      lastErr = e;
      if (shouldRetry(e) && i < retries - 1) { await sleep(delay * (i + 1)); continue; }
      if (i === retries - 1) break;
      throw e;
    }
  }
  throw lastErr;
}
function sleep(ms) { return new Promise(r => setTimeout(r, ms)); }
module.exports = { retryRequest, newSessionId, shouldRetry };
